// _oasis-engine@0.1.30@oasis-engine/dist/miniprogram.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var miniprogram = require('@oasis-engine/core/dist/miniprogram');
var miniprogram$1 = require('@oasis-engine/loader/dist/miniprogram');
var miniprogram$2 = require('@oasis-engine/math/dist/miniprogram');
var miniprogram$3 = require('@oasis-engine/rhi-webgl/dist/miniprogram');

miniprogram$1.Parser.registerComponents("o3", {
  GLTFModel: miniprogram$1.GLTFModel,
  SpriteRenderer: miniprogram.SpriteRenderer,
  PointLight: miniprogram.PointLight,
  AmbientLight: miniprogram.AmbientLight,
  DirectLight: miniprogram.DirectLight,
  EnvironmentMapLight: miniprogram.EnvironmentMapLight,
  Particle: miniprogram.Particle,
  SkyBox: miniprogram.SkyBox,
  BoxCollider: miniprogram.BoxCollider,
  GeometryRenderer: miniprogram.GeometryRenderer,
  Camera: miniprogram.Camera,
  Component: miniprogram.Component,
  SphereCollider: miniprogram.SphereCollider,
  PlaneProbe: miniprogram.PlaneProbe,
  Model: miniprogram.Model
}); //@ts-ignore

var version = "0.1.29";
console.log("oasis engine version: ".concat(version));

Object.keys(miniprogram).forEach(function (k) {
  if (k !== 'default') Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return miniprogram[k];
    }
  });
});
Object.keys(miniprogram$1).forEach(function (k) {
  if (k !== 'default') Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return miniprogram$1[k];
    }
  });
});
Object.keys(miniprogram$2).forEach(function (k) {
  if (k !== 'default') Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return miniprogram$2[k];
    }
  });
});
Object.keys(miniprogram$3).forEach(function (k) {
  if (k !== 'default') Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return miniprogram$3[k];
    }
  });
});
exports.version = version;
