// _alicov@2.11.0@alicov/lib/git.js
'use strict';

var exec = require('./exec');

exports.getGitProjectInfo = function() {
  // http://git-scm.com/book/ch2-3.html
  // use git latest commit hash
  // git log -n 1 --pretty=format:"%H"

  // https://gitlab.com/gitlab-org/gitlab-ci-multi-runner/blob/ebce92c738ce53c735e3c9124b29bbf24ea939ab/shells/bash.go#L84
  // CI_BUILD_REF=d41e95aa1312db06ffc843c1d7fa29988a6471b6
  // CI_BUILD_REF_NAME=milestone-1
  // CI_BUILD_REPO=
  // 从 gitlab-ci 环境变量读取

  var hash = process.env.CI_BUILD_REF || exec('git log -n 1 --pretty=format:"%H"');
  var comment = exec('git log -n 1 --pretty=format:"%B"');
  var committerEmail = exec('git log -n 1 --pretty=format:"%ce"');
  var committerName = exec('git log -n 1 --pretty=format:"%cn"');
  var giturl = process.env.CI_BUILD_REPO || exec('git config --get remote.origin.url');
  if (giturl) {
    // cise 会使用 mirror url
    giturl = giturl.replace('gitlab-mirror.alibaba-inc.com', 'gitlab.alibaba-inc.com');
  }
  var branch = process.env.CI_BUILD_REF_NAME ||  exec('git branch | awk \'/^\*/{print $2}\'');

  return {
    giturl: giturl,
    hash: hash,
    branch: branch,
    email: committerEmail,
    author: committerName,
    comment: comment,
    buildUrl: process.env.CI_BUILD_URL || '',
  };
};
